### ISSUE TITLE
Lambda layer to slice input not being applied

### ISSUE THREAD

**Issue by danielhers** [fchollet/keras#3673]
This is not the same as #1747 - the `output_shape` of my Lambda layer is fine, but during training the layer seems to be leaving the input unchanged: https://github.com/danielhers/ucca/blob/lstm/parsing/classifiers/nn/lstm.py#L124
This network is supposed to be slicing the inputs to `(*, self._output_max_len, *)` (first dimension is the batch size, of course, and the last varies per input but it should be OK for concatenation).
When running it, I am getting the following:

    ValueError: all the input array dimensions except for the concatenation axis must match exactly
    Apply node that caused the error: Join(TensorConstant{2}, Reshape{3}.0, Reshape{3}.0, Subtensor{int64::}.0)
    Toposort index: 102
    Inputs types: [TensorType(int8, scalar), TensorType(float32, 3D), TensorType(float32, 3D), TensorType(float32, 3D)]
    Inputs shapes: [(), (0, 1000, 20), (0, 1000, 100), (0, 1500, 6)]
    Inputs strides: [(), (80000, 80, 4), (400000, 400, 4), (36000, 24, 4)]
    Inputs values: [array(2, dtype=int8), array([], shape=(0, 1000, 20), dtype=float32), array([], shape=(0, 1000, 100), dtype=float32), array([], shape=(0, 1500, 6), dtype=float32)]
    Outputs clients: [[Elemwise{Composite{Switch(i0, (i1 * i2 * i3), i2)}}[(0, 2)](InplaceDimShuffle{x,x,x}.0, TensorConstant{(1, 1, 1) of 2.0}, Join.0, Elemwise{Composite{Cast{float32}(LT(i0, i1))}}[(0, 0)].0)]]

Here 1000 is `self._output_max_len`, 1500 is `self._max_len`, and I have 3 inputs to the network (batch size, self._max_len, `param.num` shown here):

    numeric: (10, 1500, 6)
    e: (10, 1500, 4)
    w: (10, 1500, 1)

So the concatenation at line 130 is throwing an error because not all inputs are the same modulo the last dimension, namely the input corresponding to the `numeric` input is still `(10, 1500, 6)`, as if the `slicer` layer did nothing. How can that be?

**Comment by farizrahman4u** [comment]
Post your code, please.

**Comment by danielhers** [comment]
Sorry, thought the link would suffice.
This is the relevant part:

        for suffix, param in self.feature_params.items():
            if config.Config().args.verbose:
                print("%s: %s" % (suffix, param))
            i = Input(shape=(self._max_len, param.num), dtype="float32" if param.numeric else "int32", name=suffix)
            inputs.append(i)
            slicer = Lambda(lambda a: a[self._input_max_len:], output_shape=(self._output_max_len, param.num))
            if param.indexed:  # w, t
                # first half is input representation encoding
                x = Lambda(lambda a: a[:self._input_max_len, 0], output_shape=(self._input_max_len,))(i)
                x = Embedding(output_dim=param.dim, input_dim=param.size, init=self._init,
                              weights=param.init, input_length=self._input_max_len, W_regularizer=self._regularizer)(x)
                rep = x#Bidirectional(LSTM(self._layer_dim, activation=self._activation, init=self._init,
                #                          return_sequences=True))(x) TODO uncomment
                # second half is indices into first half
                idx = slicer(i)
                x = merge([rep, idx], mode=lambda a: a[0][:, a[1]],
                          output_shape=(self._output_max_len, param.num, param.dim))#self._layer_dim))
                x = Reshape((self._output_max_len, param.num * param.dim))(x)#self._layer_dim))(x)
            else:  # not indexed: e, A, p, x or numeric
                x = slicer(i)  # TODO why is the shape wrong here for numeric input?
                if not param.numeric:
                    x = Embedding(output_dim=param.dim, input_dim=param.size, init=self._init, weights=param.init,
                                  input_length=self._output_max_len, W_regularizer=self._regularizer)(x)
                    x = Reshape((self._output_max_len, param.dim))(x)
            encoded.append(x)
        x = merge(encoded, mode="concat", concat_axis=-1) if len(encoded) > 1 else encoded[0]

**Comment by farizrahman4u** [comment]
This is in the same as #1747, there the slicing was correct, op just forgot to set output_shape. Here you are slicing it wrong. It should be:
```python
slicer = Lambda(lambda a: a[:, self._input_max_len:, :], output_shape=(self._output_max_len, param.num))
```
And,
```python
 x = Lambda(lambda a: a[:, :self._input_max_len, 0], output_shape=(self._input_max_len,))(i)
```

**Comment by danielhers** [comment]
Thanks for the answer.
I see, so I need to include the batch axis too!
Not sure about the first line, though - I should have explained it better:
The first `self._input_max_len` time points contain the input for my *indexed* params, and the last `self._output_max_len` time points of the same inputs contain indices into the first `self._input_max_len`.
For my *non-indexed* params, the first `self._input_max_len` time points are always zero, and the last `self._output_max_len` time points contain the actual data. I made it that way because I thought the inputs must all be the same length.
Then each output time point is supposed to look at the input received at the same time point (which for some of the params may be indices into the first `self._input_max_len` time points).

I changed these two lines to

    slicer = Lambda(lambda a: a[:, self._input_max_len:], output_shape=(self._output_max_len, param.num))

and

    x = Lambda(lambda a: a[:, :self._input_max_len, 0], output_shape=(self._input_max_len,))(i)

and now I'm getting this:

    ValueError: all the input array dimensions except for the concatenation axis must match exactly
    Apply node that caused the error: Join(TensorConstant{2}, Subtensor{::, :int64:}.0, Reshape{3}.0, Reshape{3}.0)
    Toposort index: 105
    Inputs types: [TensorType(int8, scalar), TensorType(float32, 3D), TensorType(float32, 3D), TensorType(float32, 3D)]
    Inputs shapes: [(), (10, 1000, 6), (100, 1000, 100), (40, 1000, 20)]
    Inputs strides: [(), (36000, 24, 4), (400000, 400, 4), (80000, 80, 4)]
    Inputs values: [array(2, dtype=int8), 'not shown', 'not shown', 'not shown']
    Outputs clients: [[Elemwise{Composite{Switch(i0, (i1 * i2 * i3), i2)}}[(0, 2)](InplaceDimShuffle{x,x,x}.0, TensorConstant{(1, 1, 1) of 2.0}, Join.0, Elemwise{Composite{Cast{float32}(LT(i0, i1))}}[(0, 0)].0)]]

So the batch size seems to be off. It's admittedly better than a batch size of 0, but I don't understand why it's 10, 100 and 40, respectively. I'm feeding a batch size of 10 for all inputs.

**Comment by danielhers** [comment]
I guess my problem is not with the slicing, but rather that my `e` input has an extra axis that I want to use, but the `Embedding` doesn't allow that. Namely, at each output time point I want to feed 4 separate `e` elements, embed them and then just flatten them so that the last dimension would be 80=4*20. But I can't do that because for some reason the 4 gets pushed into the batch size.

I don't want separate `Embedding` layers because all 4 elements should be embedded using the same lookup table. And I can't really do the trick I did with the `w` elements (feeding them in the first `self._input_max_len` time points and then indexing them) because they are relevant per output time point, not per input time point.

And even if I could artificially feed all possible `e` values in the first `self._input_max_len` time points and then just index into them, I don't know how to handle the fact that I have 4 separate indices into them. This is also the problem with the `x = merge([rep, idx], mode=lambda a: a[0][:, a[1]],` line: it gives an effective output shape of (100, 1000, 100) for some reason. I fixed that by changing it to `x = merge([rep, idx], mode=lambda a: a[0][:, a[1][0]],` which is fine because I really do have only one `w` index per time point right now, but if I want to have more than that (say, 4) it seems to be getting pushed into the batch axis again...

**Comment by farizrahman4u** [comment]
tl;dr The original issue with slicing using Lambda has been solved. For further help, you will have to state your problem in a much more simpler manner, along with a simplified a code snippet (not just dump code from your actual program.. for e.g.. I don't know what is your `e`, nor the elements in the `encoded` list ) which demonstrates what you are trying to do.

**Comment by danielhers** [comment]
Fair enough. Thanks a lot for your help so far!
